tests/returntypes/gradients/test_hadamard_grad.py

```diff
@@ -0,0 +1,853 @@
+        with qml.queuing.AnnotatedQueue() as q:
+            qml.expval(qml.PauliZ(0))
+
+        tape = qml.tape.QuantumScript.from_queue(q)
+        with pytest.warns(UserWarning, match="gradient of a tape with no trainable parameters"):
+            tapes, _ = qml.gradients.hadamard_grad(tape)
+        assert not tapes
+
+    def test_all_parameters_independent(self):
+        """Test that a circuit where all parameters do not affect the output"""
+        with qml.queuing.AnnotatedQueue() as q:
+            qml.RX(0.4, wires=0)
+            qml.expval(qml.PauliZ(1))
+
+        tape = qml.tape.QuantumScript.from_queue(q)
+        tapes, _ = qml.gradients.hadamard_grad(tape)
+        assert not tapes
+
+    def test_state_non_differentiable_error(self):
+        """Test error raised if attempting to differentiate with
+        respect to a state"""
+
+        with qml.queuing.AnnotatedQueue() as q:
+            qml.RX(0.543, wires=[0])
+            qml.RY(-0.654, wires=[1])
+            qml.state()
+
+        tape = qml.tape.QuantumScript.from_queue(q)
+        with pytest.raises(ValueError, match=r"return the state is not supported"):
+            qml.gradients.hadamard_grad(tape)
+
+    def test_independent_parameter(self, mocker):
+        """Test that an independent parameter is skipped
+        during the Jacobian computation."""
+        spy = mocker.spy(qml.gradients.hadamard_gradient, "_expval_hadamard_grad")
+
+        with qml.queuing.AnnotatedQueue() as q:
+            qml.RX(0.543, wires=[0])
+            qml.RY(-0.654, wires=[1])  # does not have any impact on the expval
+            qml.expval(qml.PauliZ(0))
+
+        tape = qml.tape.QuantumScript.from_queue(q)
+        dev = qml.device("default.qubit", wires=3)
+        tapes, fn = qml.gradients.hadamard_grad(tape)
+        assert len(tapes) == 1
+        assert tapes[0].batch_size is None
+
+        res = fn(dev.batch_execute(tapes))
+
+        assert isinstance(res, tuple)
+        assert len(res) == 2
+        assert res[0].shape == ()
+        assert res[1].shape == ()
+
+        # only called for parameter 0
+        assert spy.call_args[0][0:2] == (tape, [0])
+
+    @pytest.mark.autograd
+    def test_no_trainable_params_qnode_autograd(self, mocker):
+        """Test that the correct ouput and warning is generated in the absence of any trainable
+        parameters"""
+        dev = qml.device("default.qubit", wires=2)
+        spy = mocker.spy(dev, "expval")
+
+        @qml.qnode(dev, interface="autograd")
+        def circuit(weights):
+            qml.RX(weights[0], wires=0)
+            qml.RY(weights[1], wires=0)
+            return qml.expval(qml.PauliZ(0) @ qml.PauliZ(1))
+
+        weights = [0.1, 0.2]
+        with pytest.warns(UserWarning, match="gradient of a QNode with no trainable parameters"):
+            res = qml.gradients.hadamard_grad(circuit)(weights)
+
+        assert res == ()
+        spy.assert_not_called()
```

Comment:
Can we parametrize this test? It only differs in the `interface` argument for the QNode, if I'm not mistaken.

Reply:
We avoid that in general order to keep the right marker